index.js

```diff
@@ -91,44 +148,42 @@ app.get('/api/names/labels', (req, res) => {
-// about and hour faster than uploading 1112 files
-const SERVERS=[
-  "https://quick-start-xandr-videohost.s3-us-west-2.amazonaws.com/",
-  "https://crowd-source-circuit-tv.s3-us-west-1.amazonaws.com/dev_splits_complete/"
-]
-
-// round robin load balancer
-function getServerIndex(){
-// global yes indeed
-  VIDEO_SERVER_ID = (VIDEO_SERVER_ID+1)%NUMBER_SERVERS;
-  return VIDEO_SERVER_ID;
-
-}
 // Endpoint - Select next video
 app.get('/api/videos/select/username/:username', (req, res) => {
+  // Remove all videos that have been checked out for too long
+  var now = Date.now()
+  for (let i = checkedOutVideos.length - 1; i >= 0; i--) {
+    if (now - checkedOutVideos[i].timeStamp > TIMEOUT)
+      checkedOutVideos.splice(i, 1);
+  }
+
   // Declare helping variable
   var usableFileList = [];
 
-  // Find partially completed videos not seen be user
-  db.query("SELECT fileTitle FROM Videos WHERE id IN (SELECT videoid FROM Votes WHERE videoid NOT IN (SELECT videoid FROM Votes WHERE userid = (SELECT id FROM Users WHERE username = $1)) GROUP BY videoid HAVING COUNT(*) < $2);", [req.params.username, LABEL_MAX], (err, result) => {
+  // Find partially completed videos not seen be user of form {filetitle: X, count: Y}
+  db.query("SELECT V.fileTitle, COUNT(videoid) FROM Votes INNER JOIN Videos V ON V.id = videoid WHERE videoid NOT IN (SELECT videoid FROM Votes WHERE userid = (SELECT id FROM Users WHERE username = $1)) GROUP BY fileTitle HAVING COUNT(videoid) < $2 ORDER BY fileTitle;", [req.params.username, LABEL_MAX], (err, result) => {
     if (err) {
       res.status(500).json({content: `Error selecting partially voted video for user: ${req.params.username}`});
       console.log(err);
       return;
     }
     else {
-      usableFileList = result.rows.map(entry => entry.filetitle);
+      // Convert each entry from form {filetitle: X, count: Y} to form [X, X, X,... (LABEL_MAX - Y times)] and flatten the entries into a single array
+      //   this is done because Y is the number of votes in the database, and LABEL_MAX - Y is the number of times that video is allowed to be checked out without exceeding LABEL_MAX
+      usableFileList = result.rows.myFlatMap(entry => Array(LABEL_MAX - entry.count).fill(entry.filetitle));
+      // Take only the videos in usable file list which are not checked out
+      checkedOutVideos.map(entry => entry.video).forEach(video => {
+        index = usableFileList.indexOf(video);
+        if (index > -1)
+          usableFileList.splice(index, 1);
+      });
 
       // There are at least one partially complete videos unseen by the user, select one and send it
       if (usableFileList.length > 0) {
         const server = SERVERS[getServerIndex()];
         const fileid = usableFileList[Math.floor(100*Math.random() % usableFileList.length)];
+        checkedOutVideos.push({'video': fileid, 'timeStamp': now});
+        console.log(checkedOutVideos);
```

Comment:
Based on this log

![image](https://user-images.githubusercontent.com/50972261/74403779-548f2580-4ddd-11ea-990f-89e992ef6580.png)

we can see if a partially video is requested in succession of it arriving on the checked out list, it gets duplicated. they then expire on seperate times. 
maybe you could key a map on your flatten function to basically just delete the last existing copy of the filename. which would maintain only 1 of each in the list.

but other than that this console.log seems to work as expected. I refreshed the page 5 times, generated a list of checked out videos, then after 10 minutes i saw that they we removed from checked out cart, containing the new one that was created on "submit and refresh"
]